partitioner/src/main/java/io/confluent/connect/storage/partitioner/TimeBasedPartitioner.java

```diff
@@ -103,8 +137,87 @@ public String encodePartition(SinkRecord sinkRecord) {
           (Class<? extends SchemaGenerator<T>>) config.get(PartitionerConfig.SCHEMA_GENERATOR_CLASS_CONFIG);
       return generatorClass.getConstructor(Map.class).newInstance(config);
     } catch (ClassCastException | IllegalAccessException | InstantiationException | InvocationTargetException
-        | NoSuchMethodException e) {
+                 | NoSuchMethodException e) {
       throw new ConfigException("Invalid generator class: " + generatorClass);
     }
   }
+
+  public TimestampExtractor newTimestampExtractor(String extractorClassName) {
+    try {
+      switch (extractorClassName) {
+        case "Wallclock":
+        case "Record":
+        case "RecordField":
+          extractorClassName = "io.confluent.connect.storage.partitioner.TimeBasedPartitioner$" + extractorClassName
+                                   + "TimestampExtractor";
+          break;
+        default:
+      }
+      Class<?> klass = Class.forName(extractorClassName);
+      if (!TimestampExtractor.class.isAssignableFrom(klass)) {
+        throw new ConnectException("Class " + extractorClassName + " does not implement TimestampExtractor");
+      }
+      return (TimestampExtractor) klass.newInstance();
+    } catch (ClassNotFoundException | ClassCastException | IllegalAccessException | InstantiationException e) {
+      throw new ConfigException("Invalid timestamp extractor: " + extractorClassName, e);
+    }
+  }
+
+  public static class WallclockTimestampExtractor implements TimestampExtractor {
+    @Override
+    public void configure(Map<String, Object> config) {}
+
+    @Override
+    public Long extract(ConnectRecord<?> record) {
+      return Time.SYSTEM.milliseconds();
+    }
+  }
+
+  public static class RecordTimestampExtractor implements TimestampExtractor {
+    @Override
+    public void configure(Map<String, Object> config) {}
+
+    @Override
+    public Long extract(ConnectRecord<?> record) {
+      return record.timestamp();
+    }
+  }
+
+  public static class RecordFieldTimestampExtractor implements TimestampExtractor {
+    private String fieldName;
+
+    @Override
+    public void configure(Map<String, Object> config) {
+      fieldName = (String) config.get(PartitionerConfig.TIMESTAMP_FIELD_NAME_CONFIG);
+    }
+
+    @Override
+    public Long extract(ConnectRecord<?> record) {
+      Object value = record.value();
+      Schema valueSchema = record.valueSchema();
+      if (value instanceof Struct) {
+        Struct struct = (Struct) value;
+        Object timestampValue = struct.get(fieldName);
+        Type type = valueSchema.field(fieldName).schema().type();
+        switch (type) {
+          case INT8:
+          case INT16:
+          case INT32:
+          case INT64:
+            Number timestamp = (Number) timestampValue;
+            return timestamp.longValue();
+          case STRING:
+            // Should we support parsing from string?
+          case BYTES:
+            // Do we support serialized timestamps?
```

Comment:
Serialized according to what format? I think we can likely just ignore `BYTES`. The only case I can think of that makes sense here is if it's actually a `BigDecimal`.